core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp

```diff
@@ -599,40 +599,27 @@ class ParallelFor<FunctorType, Kokkos::TeamPolicy<Properties...>,
     // Maximum active teams possible.
     int max_active_teams = OpenMPTargetExec::MAX_ACTIVE_THREADS / team_size;
 
-    int* lock_array = OpenMPTargetExec::get_lock_array(max_active_teams);
-
-#pragma omp target teams distribute map(to                   \
-                                        : a_functor)         \
-    is_device_ptr(scratch_ptr, lock_array) num_teams(nteams) \
-        thread_limit(team_size)
-    for (int i = 0; i < league_size; i++) {
-      int shmem_block_index = -1, lock_team = 99999, iter = -1;
-      iter = (omp_get_team_num() % max_active_teams);
-
-      // Loop as long as a shmem_block_index is not found.
-      while (shmem_block_index == -1) {
-        // Try and acquire a lock on the index.
-        lock_team = atomic_compare_exchange(&lock_array[iter], 0, 1);
-
-        // If lock is acquired assign it to the block index.
-        // lock_team = 0, implies atomic_compare_exchange is successfull.
-        if (lock_team == 0)
-          shmem_block_index = iter;
-        else
-          iter = ++iter % max_active_teams;
-      }
-
-#pragma omp parallel num_threads(team_size)
-      {
-        typename Policy::member_type team(
-            i, league_size, team_size, vector_length, scratch_ptr,
-            shmem_block_index, shmem_size_L0, shmem_size_L1);
+// Performing our own scheduling of teams to avoid separation of code between
+// teams-distribute and parallel. Gave a 2x performance boost in test cases.
+#pragma omp target teams num_teams(nteams) thread_limit(team_size) \
+    map(to                                                         \
+        : a_functor) is_device_ptr(scratch_ptr)
+#pragma omp parallel
+    {
+      const int blockIdx = omp_get_team_num();
+      const int gridDim  = omp_get_num_teams();
+
+      // Iterate through the number of teams until league_size and assign the
+      // league_id accordingly
+      for (int league_id = blockIdx, track = 0; league_id < league_size;
```

Comment:
Does that loop always works? I feel that this works only **if** `league_size < OpenMPTargetExec::MAX_ACTIVE_TEAMS` and `num_teams = nteams` which does the standard does not guarantee. The comment indicates that you want to iterate thought the number of teams but `league_size` is an upper of that number, right?

Reply:
@Rombur - The standard guarantees that , the number of teams specified in the `num_teams` clause is always less than or equal to the maximum concurrently running teams. So I think the loop would work by looping around the `gridDim` generated. I tried it with a test case where it successfully works when the `league_size > max_active_teams`.